s2: non Indexed Seek may seek to the wrong offset

Hi!

I have noticed an issue where the non-indexed Seek code may skip to the wrong offset depending on what parameters are used.  I'm not quite sure what is wrong yet, but I was able to work out a test case that shows the issue well.

This writes 1,000,000 25 byte records, then attempts to seek through them using different skip lengths.  The part that made this hard to track down is that for small skip sizes it works properly, but large sizes return invalid data, or try seeking too far and trigger `unexpected EOF`.

```
$ go test -v      
=== RUN   TestSeeking
=== RUN   TestSeeking/skip=100
=== RUN   TestSeeking/skip=1000
=== RUN   TestSeeking/skip=10000
=== RUN   TestSeeking/skip=20000
=== RUN   TestSeeking/skip=100000
    main_test.go:64: Expected "Item 0000000000000200000\n", got "tem 0000000000000241943\nI"
=== RUN   TestSeeking/skip=200000
    main_test.go:64: Expected "Item 0000000000000400000\n", got "m 0000000000000525829\nIte"
=== RUN   TestSeeking/skip=400000
    main_test.go:56: Failed to seek: unexpected EOF
--- FAIL: TestSeeking (0.21s)
    --- PASS: TestSeeking/skip=100 (0.01s)
    --- PASS: TestSeeking/skip=1000 (0.01s)
    --- PASS: TestSeeking/skip=10000 (0.01s)
    --- PASS: TestSeeking/skip=20000 (0.01s)
    --- FAIL: TestSeeking/skip=100000 (0.00s)
    --- FAIL: TestSeeking/skip=200000 (0.00s)
    --- FAIL: TestSeeking/skip=400000 (0.00s)
```


```
import (
        "fmt"
        "io"
        "os"
        "path/filepath"
        "testing"

        "github.com/klauspost/compress/s2"
)

func TestSeeking(t *testing.T) {
        d := t.TempDir()
        fn := filepath.Join(d, "test.s2")

        f, err := os.Create(fn)
        if err != nil {
                t.Fatal(err)
        }

        enc := s2.NewWriter(f)

        //24 bytes per item plus \n = 25 bytes per record
        for i := 0; i < 1_000_000; i++ {
                fmt.Fprintf(enc, "Item %019d\n", i)
        }

        err = enc.Close()
        if err != nil {
                t.Fatal(err)
        }

        err = f.Close()
        if err != nil {
                t.Fatal(err)
        }

        for _, skip := range []int{100, 1_000, 10_000, 20_000, 100_000, 200_000, 400_000} {
                t.Run(fmt.Sprintf("skip=%d", skip), func(t *testing.T) {
                        f, err = os.Open(fn)
                        if err != nil {
                                t.Fatal(err)
                        }
                        defer f.Close()
                        dec := s2.NewReader(f)
                        seeker := s2.ReadSeeker{
                                Reader: dec,
                        }
                        buf := make([]byte, 25)
                        for rec := 0; rec < 1_000_000; rec += skip {
                                offset := int64(rec * 25)
                                //t.Logf("Reading record %d", rec)
                                _, err := seeker.Seek(offset, io.SeekStart)
                                if err != nil {
                                        t.Fatalf("Failed to seek: %v", err)
                                }
                                _, err = dec.Read(buf)
                                if err != nil {
                                        t.Fatalf("Failed to seek: %v", err)
                                }
                                expected := fmt.Sprintf("Item %019d\n", rec)
                                if string(buf) != expected {
                                        t.Fatalf("Expected %q, got %q", expected, buf)
                                }
                        }
                })
        }

}
```

--- a/s2/decode.go
+++ b/s2/decode.go
@@ -791,6 +791,7 @@ func (r *Reader) Skip(n int64) error {
 			} else {
 				// Skip block completely
 				n -= int64(dLen)
+				r.blockStart += int64(dLen)
 				dLen = 0
 			}
 			r.i, r.j = 0, dLen
@@ -921,6 +922,15 @@ func (r *Reader) ReadSeeker(random bool, index []byte) (*ReadSeeker, error) {
 	err = r.index.LoadStream(rs)
 	if err != nil {
 		if err == ErrUnsupported {
+			// If we don't require random seeking, reset input and return.
+			if !random {
+				_, err = rs.Seek(pos, io.SeekStart)
+				if err != nil {
+					return nil, ErrCantSeek{Reason: "resetting stream returned: " + err.Error()}
+				}
+				r.index = nil
+				return &ReadSeeker{Reader: r}, nil
+			}
 			return nil, ErrCantSeek{Reason: "input stream does not contain an index"}
 		}
 		return nil, ErrCantSeek{Reason: "reading index returned: " + err.Error()}

--- a/s2/index_test.go
+++ b/s2/index_test.go
@@ -112,6 +112,7 @@ func TestSeeking(t *testing.T) {
 		nElems = 100_000
 		testSizes = []int{100, 1_000, 10_000, 20_000}
 	}
+	testSizes = append(testSizes, nElems-1)
 	//24 bytes per item plus \n = 25 bytes per record
 	for i := 0; i < nElems; i++ {
 		fmt.Fprintf(enc, "Item %019d\n", i)
@@ -130,7 +131,7 @@ func TestSeeking(t *testing.T) {
 				t.Fatal(err)
 			}
 			buf := make([]byte, 25)
-			for rec := 0; rec < 1_000_000; rec += skip {
+			for rec := 0; rec < nElems; rec += skip {
 				offset := int64(rec * 25)
 				//t.Logf("Reading record %d", rec)
 				_, err := seeker.Seek(offset, io.SeekStart)
@@ -154,7 +155,7 @@ func TestSeeking(t *testing.T) {
 				t.Fatal(err)
 			}
 			buf := make([]byte, 25)
-			for rec := 0; rec < 1_000_000; rec += skip {
+			for rec := 0; rec < nElems; rec += skip {
 				offset := int64(rec * 25)
 				//t.Logf("Reading record %d", rec)
 				_, err := seeker.Seek(offset, io.SeekStart)
@@ -178,7 +179,7 @@ func TestSeeking(t *testing.T) {
 				t.Fatal(err)
 			}
 			buf := make([]byte, 25)
-			for rec := 0; rec < 1_000_000; rec += skip {
+			for rec := 0; rec < nElems; rec += skip {
 				offset := int64(rec * 25)
 				//t.Logf("Reading record %d", rec)
 				_, err := seeker.Seek(offset, io.SeekStart)
@@ -203,7 +204,7 @@ func TestSeeking(t *testing.T) {
 				t.Fatal(err)
 			}
 			buf := make([]byte, 25)
-			for rec := 0; rec < 1_000_000; rec += skip {
+			for rec := 0; rec < nElems; rec += skip {
 				offset := int64(rec * 25)
 				//t.Logf("Reading record %d", rec)
 				_, err := seeker.Seek(offset, io.SeekStart)

--- a/s2/index_test.go
+++ b/s2/index_test.go
@@ -7,6 +7,7 @@ import (
 	"io/ioutil"
 	"math/rand"
 	"sync"
+	"testing"
 
 	"github.com/klauspost/compress/s2"
 )
@@ -98,4 +99,126 @@ func ExampleIndex_Load() {
 	//Successfully skipped forward to 3888885
 	//Successfully skipped forward to 4444440
 	//Successfully skipped forward to 4999995
+}
+
+func TestSeeking(t *testing.T) {
+	compressed := bytes.Buffer{}
+
+	// Use small blocks so there are plenty of them.
+	enc := s2.NewWriter(&compressed, s2.WriterBlockSize(16<<10))
+	var nElems = 1_000_000
+	var testSizes = []int{100, 1_000, 10_000, 20_000, 100_000, 200_000, 400_000}
+	if testing.Short() {
+		nElems = 100_000
+		testSizes = []int{100, 1_000, 10_000, 20_000}
+	}
+	//24 bytes per item plus \n = 25 bytes per record
+	for i := 0; i < nElems; i++ {
+		fmt.Fprintf(enc, "Item %019d\n", i)
+	}
+
+	index, err := enc.CloseIndex()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	for _, skip := range testSizes {
+		t.Run(fmt.Sprintf("noSeekSkip=%d", skip), func(t *testing.T) {
+			dec := s2.NewReader(io.NopCloser(bytes.NewReader(compressed.Bytes())))
+			seeker, err := dec.ReadSeeker(false, nil)
+			if err != nil {
+				t.Fatal(err)
+			}
+			buf := make([]byte, 25)
+			for rec := 0; rec < 1_000_000; rec += skip {
+				offset := int64(rec * 25)
+				//t.Logf("Reading record %d", rec)
+				_, err := seeker.Seek(offset, io.SeekStart)
+				if err != nil {
+					t.Fatalf("Failed to seek: %v", err)
+				}
+				_, err = io.ReadFull(dec, buf)
+				if err != nil {
+					t.Fatalf("Failed to seek: %v", err)
+				}
+				expected := fmt.Sprintf("Item %019d\n", rec)
+				if string(buf) != expected {
+					t.Fatalf("Expected %q, got %q", expected, buf)
+				}
+			}
+		})
+		t.Run(fmt.Sprintf("seekSkip=%d", skip), func(t *testing.T) {
+			dec := s2.NewReader(io.ReadSeeker(bytes.NewReader(compressed.Bytes())))
+			seeker, err := dec.ReadSeeker(false, nil)
+			if err != nil {
+				t.Fatal(err)
+			}
+			buf := make([]byte, 25)
+			for rec := 0; rec < 1_000_000; rec += skip {
+				offset := int64(rec * 25)
+				//t.Logf("Reading record %d", rec)
+				_, err := seeker.Seek(offset, io.SeekStart)
+				if err != nil {
+					t.Fatalf("Failed to seek: %v", err)
+				}
+				_, err = io.ReadFull(dec, buf)
+				if err != nil {
+					t.Fatalf("Failed to seek: %v", err)
+				}
+				expected := fmt.Sprintf("Item %019d\n", rec)
+				if string(buf) != expected {
+					t.Fatalf("Expected %q, got %q", expected, buf)
+				}
+			}
+		})
+		t.Run(fmt.Sprintf("noSeekIndexSkip=%d", skip), func(t *testing.T) {
+			dec := s2.NewReader(io.NopCloser(bytes.NewReader(compressed.Bytes())))
+			seeker, err := dec.ReadSeeker(false, index)
+			if err != nil {
+				t.Fatal(err)
+			}
+			buf := make([]byte, 25)
+			for rec := 0; rec < 1_000_000; rec += skip {
+				offset := int64(rec * 25)
+				//t.Logf("Reading record %d", rec)
+				_, err := seeker.Seek(offset, io.SeekStart)
+				if err != nil {
+					t.Fatalf("Failed to seek: %v", err)
+				}
+				_, err = io.ReadFull(dec, buf)
+				if err != nil {
+					t.Fatalf("Failed to seek: %v", err)
+				}
+				expected := fmt.Sprintf("Item %019d\n", rec)
+				if string(buf) != expected {
+					t.Fatalf("Expected %q, got %q", expected, buf)
+				}
+			}
+		})
+		t.Run(fmt.Sprintf("seekIndexSkip=%d", skip), func(t *testing.T) {
+			dec := s2.NewReader(io.ReadSeeker(bytes.NewReader(compressed.Bytes())))
+
+			seeker, err := dec.ReadSeeker(false, index)
+			if err != nil {
+				t.Fatal(err)
+			}
+			buf := make([]byte, 25)
+			for rec := 0; rec < 1_000_000; rec += skip {
+				offset := int64(rec * 25)
+				//t.Logf("Reading record %d", rec)
+				_, err := seeker.Seek(offset, io.SeekStart)
+				if err != nil {
+					t.Fatalf("Failed to seek: %v", err)
+				}
+				_, err = io.ReadFull(dec, buf)
+				if err != nil {
+					t.Fatalf("Failed to seek: %v", err)
+				}
+				expected := fmt.Sprintf("Item %019d\n", rec)
+				if string(buf) != expected {
+					t.Fatalf("Expected %q, got %q", expected, buf)
+				}
+			}
+		})
+	}
 }

